client/my-sites/stats/features/modules/stats-locations/stats-locations.tsx

```diff
@@ -101,32 +105,99 @@ const StatsLocations: React.FC< StatsModuleLocationsProps > = ( {
 
 	// Main location data query
 	const {
-		data: locationsViewsData = [],
+		data: locationsViewsData,
 		isLoading: isRequestingData,
 		isError,
 	} = useLocationViewsQuery< StatsLocationViewsData >( siteId, geoMode, query, countryFilter, {
 		enabled: ! shouldGate && supportsLocationsStatsFeature,
 	} );
 
+	const normalizedLocationsViewsData = useMemo( () => {
+		if ( isRequestingData || ! locationsViewsData ) {
+			return [];
+		}
+
+		const normalizedStats = normalizers.statsCountryViews(
+			locationsViewsData as StatsLocationViewsData,
+			query
+		);
+
+		if ( ! Array.isArray( normalizedStats ) ) {
+			return [];
+		}
+
+		return query?.max ? normalizedStats.slice( 0, query.max ) : normalizedStats;
+	}, [ locationsViewsData, query, isRequestingData ] );
+
 	// The legacy endpoint that only supports countries (not regions or cities)
 	// will be used when the new Locations Stats feature is not available.
 	const legacyCountriesViewsData = useSelector( ( state ) =>
 		getSiteStatsNormalizedData( state, siteId, statType, query )
 	) as [ id: number, label: string ];
 
-	const data = supportsLocationsStatsFeature ? locationsViewsData : legacyCountriesViewsData;
+	const data = supportsLocationsStatsFeature
+		? normalizedLocationsViewsData
+		: legacyCountriesViewsData;
 
 	// Only fetch separate countries list if we're not already in country tab
 	// This is to avoid fetching the same data twice.
-	const { data: countriesList = [] } = useLocationViewsQuery< StatsLocationViewsData >(
-		siteId,
-		'country',
-		query,
-		null,
-		{
+	const { data: countriesList, isLoading: isRequestingCountriesList } =
+		useLocationViewsQuery< StatsLocationViewsData >( siteId, 'country', query, null, {
 			enabled: ! shouldGate && supportsLocationsStatsFeature && geoMode !== 'country',
+		} );
+
+	const normalizedCountriesList = useMemo( () => {
+		if ( isRequestingCountriesList || ! countriesList ) {
+			return [];
 		}
-	);
+
+		const normalizedStats = normalizers.statsCountryViews(
+			countriesList as StatsLocationViewsData,
+			query
+		);
+
+		if ( ! Array.isArray( normalizedStats ) ) {
+			return [];
+		}
+
+		return normalizedStats;
+	}, [ countriesList, query, isRequestingCountriesList ] );
+
+	// Data is fetched from three ways but only one is displayed, we use that for downloading as CSV.
+	// If supportsLocationsStatsFeature is false, we use the legacy endpoint.
+	// If geoMode is country, we use the countriesList.
+	// Otherwise, we use the locationsViewsData.
+	const dataToDispatch = useMemo( () => {
```

Comment:
If `dataToDispatch` is only used in useEffect, then I guess we could move it into that